<h1><center><b>DATA PREPARATION</b></center></h1>

#### We will take a classic dataset that every data scientist has seen hundreds of times and import the Python scikit-learn module in order to examine this functionality: The Iris data set used by British scientist Ronald Fisher in his 1936 article "The use of multiple measurements in taxonomic problems"

In [8]:
import pandas as pd
from sklearn.datasets import load_iris

iris = load_iris()
iris_df = pd.DataFrame(data=iris['data'], columns=iris['feature_names'])

## DataFrame.shape
The variable iris_df has the Iris Dataset imported into it. It would be helpful to be aware of how many datapoints we have and how big the dataset is overall before delving into the data. Examining the amount of data we are working with is helpful.

In [9]:
iris_df.shape

(150, 4)

As a result, there are 150 rows and 4 columns of data. One datapoint is represented by each row, and one feature is represented by each column in the data frame. Consequently, there are 150 datapoints with a total of 4 characteristics.

# DataFrame.columns

Let's get to the four data columns now. What precisely do they each stand for? The columns attribute will provide the dataframe's column names.

In [10]:
iris_df.columns

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)'],
      dtype='object')

There are four(4) columns, as we can see. The only information provided by the columns attribute is the name of the columns. When trying to determine what features a dataset has, this attribute becomes crucial.

# DataFrame.info

We can learn something about the dataset from the amount of data (provided by the shape parameter) and the names of the features or columns (provided by the columns attribute). We should now delve more deeply into the dataset. This is where the DataFrame.info() function comes in handy.

In [11]:
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


From here, we can draw a few results:

Each column's DataType is as follows: All of the information in this dataset is kept as 64-bit floating-point values.
Non-Null values in number: Data preparation involves dealing with null values, which is a crucial step. It will be addressed in the notebook at a later time.

# DataFrame.describe()

Let's say our dataset has a lot of numerical information. On each of the columns separately, one can perform univariate statistical computations such as the mean, median, quartiles, and so forth. We may get a statistical overview of a dataset's numerical columns using the DataFrame.describe() function.

In [12]:
iris_df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)  
count        150.000000  
mean           1.199333  
std            0.762238  
min            0.100000  
25%            0.300000  
50%            1.300000  
75%            1.800000  
max            2.500000

The output above displays the total number of data points, mean, standard deviation, minimum, lower quartile (25%) median (50%) upper quartile (75%) and maximum value for each column.

# DataFrame.head

We now have a high level view of the dataset thanks to all the aforementioned functions and characteristics. We are aware of the total number of data points, the total number of features, the data type of each feature, and the total number of non-null values for each feature.

It's time to examine the facts directly now. Let's take a look at our DataFrame's first few rows (or datapoints):

In [13]:
iris_df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

We can see five (5) elements from the dataset as the output in this case. These are the first five rows, as can be seen by looking at the index on the left.

## Exercise(a):

In [14]:
iris_df[:10]

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2
5                5.4               3.9                1.7               0.4
6                4.6               3.4                1.4               0.3
7                5.0               3.4                1.5               0.2
8                4.4               2.9                1.4               0.2
9                4.9               3.1                1.5               0.1

# DataFrame.tail

The data can also be viewed from the end rather than the beginning. DataFrame.tail, which delivers a DataFrame's final five rows, is the opposite of DataFrame.head.

In [15]:
iris_df.tail()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

####  Missing Data
Let's examine the missing data. When a value is not recorded in some of the columns, missing data occurs.

Consider the following scenario: A person who is aware of their weight declines to respond to a survey's weight question. The weight value for that particular person will thereafter be absent.

Missing values are common in datasets from the real world.

#### How missing data is handled by Pandas


Pandas has two methods for handling missing values. The first is NaN, or Not a Number, which you have already encountered in earlier sections. This is a unique number that is exclusively used to signify missing floating-point values and is a component of the IEEE floating-point specification.

Apart from floats, pandas employs the Python None object for missing values. While it may be perplexing to see two different types of values that essentially say the same thing, this design decision was made for strong programmatic reasons, and in reality, it allows pandas to provide a good compromise in the great majority of circumstances. Despite this, there are limitations associated with the use of None and NaN that you should be aware of.

## None: non-float missing data
None cannot be used in NumPy and pandas arrays that are not of the data type "object" because it is a Python construct. Keep in mind that pandas data structures and NumPy arrays can only hold one type of data. They have incredible capability for handling massive amounts of data and performing computational tasks because of this, but it also restricts their versatility. Such arrays must be converted to their "lowest common denominator," or to a data type that can hold all the elements of the array. If None appears in the array, Python objects are being used.

Consider the example array below (notice the dtype for it) to see this in action:

In [16]:
import numpy as np

example1 = np.array([2, None, 6, 8])
example1

array([2, None, 6, 8], dtype=object)

Upcast data types actually have two unintended consequences. First, rather than using compiled NumPy code, operations will be performed at the level of interpreted Python code. This basically means that any actions on Series or DataFrames that include None will take longer. Although you probably wouldn't notice this speed hit, it might become a problem for large datasets.

The first side effect feeds into the second. Arrays that contain a None value will typically result in an error when using NumPy/pandas aggregations like sum() or min() since None effectively drags Series or DataFrames back into the world of plain Python.

In [17]:
example1.sum()

TypeError: unsupported operand type(s) for +: 'int' and 'NoneType'

## NaN: missing float values
For its quick, vectorized operations and ufuncs, NaN is supported by NumPy (and hence, pandas) as opposed to None. The bad news is that NaN always yields NaN when an operation is performed on it. For instance

In [18]:
np.nan + 1

nan

In [19]:
np.nan * 0

nan

The good news is that aggregations performed on arrays containing NaN do not cause errors. The bad news is that not all of the outcomes are helpful:

In [20]:
example2 = np.array([2, np.nan, 6, 8]) 
example2.sum(), example2.min(), example2.max()

(nan, nan, nan)

## Exercise(b):

In [21]:
# What happens if you add np.nan and None together?
np.nan + None

TypeError: unsupported operand type(s) for +: 'float' and 'NoneType'

## NaN and None: null values in pandas
Despite the fact that NaN and None can exhibit slightly distinct behaviors, pandas is designed to handle them interchangeably. Consider a series of integers to understand what this mean.

In [22]:
int_series = pd.Series([1, 2, 3], dtype=int)
int_series

0    1
1    2
2    3
dtype: int32

## Exercise (c):

In [23]:
# Now set an element of int_series equal to None.
# How does that element show up in the Series?
# What is the dtype of the Series?

int_series = pd.Series([None])
int_series

0    None
dtype: object

## Detecting null values
Knowing the significance of missing values, we must first find them in our dataset before addressing them. Your primary methods for identifying null data are isnull() and notnull(). Both give you data over Boolean masks.

In [24]:
example3 = pd.Series([0, np.nan, '', None])
example3.isnull()

0    False
1     True
2    False
3     True
dtype: bool

Take a good look at the result. What about it surprises you? Despite the fact that 0 is an arithmetic null, pandas recognizes it as a valid integer. is a little more subdued than. Although we used it in Section 1 to represent an empty string value, it is still a string object and does not, in the eyes of pandas, represent null.

Let's turn this around now and apply these techniques more practically, as you would. When working with isolated missing (or present) data, you can utilize Boolean masks directly as a Series or DataFrame index.

Simply add the mask created by the isnull() method to get the total number of missing values.

In [25]:
example3.isnull().sum()

2

## Exercise (d):

In [26]:
# Try running example3[example3.notnull()].
# Before you do so, what do you expect to see?

example3[example3.notnull()]

0    0
2     
dtype: object

### Dealing with missing data
Missing data must be handled by machine learning models. Therefore, we must deal with these missing values before supplying the data to the model.

There are various compromises that can impact your final analysis and the results in the actual world depending on how missing data is handled.

The two main methods for handling missing data are:
1. Drop the row containing the missing value
2. Replace the missing value with some other value

### Dropping null values
The quantity of information we give to our model directly impacts how well it performs. By eliminating null values, we are minimizing the amount of datapoints and, as a result, the dataset size. So, when the dataset is quite huge, it is advised to delete rows containing null values.

Another situation can be if a certain row or column has a large number of missing values. Then, they might have been dropped because they wouldn't have added anything to our study because most of the data for that row or column is absent.

In [27]:
example3 = example3.dropna()
example3

0    0
2     
dtype: object

Keep in mind that this should resemble your example3[example3.notnull()] output. The distinction is that dropna has eliminated those missing values from the Series example3 rather than only indexing on the masked data.

DataFrames have additional possibilities for eliminating data because they have two dimensions.

In [28]:
example4 = pd.DataFrame([[1,      np.nan, 7], 
                         [2,      5,      8], 
                         [np.nan, 6,      9]])
example4

0    1  2
0  1.0  NaN  7
1  2.0  5.0  8
2  NaN  6.0  9

You must drop entire rows or columns in order to remove individual values from a DataFrame. Pandas gives you options for both because you might wish to do one or the other depending on what you are doing. You are more likely to remove rows of data than columns because in data science, columns often represent variables and rows represent observations; the default setting for dropna() is to remove all rows that include any null values.

In [29]:
example4.dropna()

0    1  2
1  2.0  5.0  8

If necessary, you can drop NA values from columns. Use axis=1 to do so:

In [30]:

example4.dropna(axis='columns')

2
0  7
1  8
2  9

Notice that this can drop a lot of data that you might want to keep, particularly in smaller datasets. What if you just want to drop rows or columns that contain several or even just all null values? You specify those setting in dropna with the how and thresh parameters.

By default, how='any' (if you would like to check for yourself or see what other parameters the method has, run example4.dropna? in a code cell). You could alternatively specify how='all' so as to drop only rows or columns that contain all null values. Let's expand our example DataFrame to see this in action in the next exercise.

In [31]:
example4[3] = np.nan
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

## Exercise(e):

In [32]:
example4.dropna(axis='rows', thresh=3)

0    1  2   3
1  2.0  5.0  8 NaN

### Filling null values
Sometimes it makes sense to replace missing values with potential valid ones. Several methods can be used to fill in null data. The first involves approximating the missing values using domain knowledge (knowledge of the subject on which the dataset is based).

To do this in place, you could use isnull, but that might be time-consuming, especially if you have several values to fill. Due to the prevalence of this activity in data science, pandas offers fillna, which returns a duplicate of the Series or DataFrame with the missing values filled with your preferred values. To see this in action, let's make another example Series.

### Categorical Data(Non-numeric)
Let's start by thinking about non-numeric data. We have columns containing categorical data in datasets. Including gender, true or false, etc.

In the majority of these situations, we use the column's mode to fill in any missing values. Let's say that out of 100 data points, 90 have indicated True, 8 have indicated False, and 2 have not filled out. Then, taking into account the entire column, we can will the 2 with True.

Again, this is where we can apply our domain knowledge. Let's look at a filling with the mode example.

In [33]:
fill_with_mode = pd.DataFrame([[1,2,"True"],
                               [3,4,None],
                               [5,6,"False"],
                               [7,8,"True"],
                               [9,10,"True"]])

fill_with_mode

0   1      2
0  1   2   True
1  3   4   None
2  5   6  False
3  7   8   True
4  9  10   True

Let's find the mode first, then fill the None value with it.

In [34]:
fill_with_mode[2].value_counts()

True     3
False    1
Name: 2, dtype: int64

We shall therefore swap None with True.

In [35]:
fill_with_mode[2].fillna('True',inplace=True)
fill_with_mode

0   1      2
0  1   2   True
1  3   4   True
2  5   6  False
3  7   8   True
4  9  10   True

### Numeric Data
Now, coming to numeric data. Here, we have a two common ways of replacing missing values:

1. Replace with Median of the row
2. Replace with Mean of the row

In cases of skewed data caused by outliers, we substitute the median. The median is resilient to outliers, which explains this.

When the data has been standardized, the mean can be used because the mean and median would then be quite near.

Let's begin by using a column that has a regularly distributed value to replace any missing values.

In [36]:
fill_with_mean = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [np.nan,4,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  NaN  4  5
3  1.0  6  7
4  2.0  8  9

The mean of the column is

In [37]:
np.mean(fill_with_mean[0])

0.0

Filling with mean

In [38]:
fill_with_mean[0].fillna(np.mean(fill_with_mean[0]),inplace=True)
fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  0.0  4  5
3  1.0  6  7
4  2.0  8  9

As we can see, the mean has taken the place of the missing value.

Let's try a different dataframe and replace the None values with the column's median this time.

In [39]:
fill_with_median = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [0,np.nan,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  NaN  5
3  1  6.0  7
4  2  8.0  9

The median of the second column is

In [40]:
fill_with_median[1].median()

4.0

Filling with median

In [41]:
fill_with_median[1].fillna(fill_with_median[1].median(),inplace=True)
fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  4.0  5
3  1  6.0  7
4  2  8.0  9

As can be seen, the column's median has taken the place of the NaN value.

In [42]:
example5 = pd.Series([1, np.nan, 2, None, 3], index=list('abcde'))
example5

a    1.0
b    NaN
c    2.0
d    NaN
e    3.0
dtype: float64

You can fill all of the null entries with a single value, such as 0:

In [43]:
example5.fillna(0)

a    1.0
b    0.0
c    2.0
d    0.0
e    3.0
dtype: float64

### Exercise (f):

In [44]:
example5.fillna(method='ffill')

a    1.0
b    1.0
c    2.0
d    2.0
e    3.0
dtype: float64

In [45]:
example5.fillna(method='bfill')

a    1.0
b    2.0
c    2.0
d    3.0
e    3.0
dtype: float64

In [46]:
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

In [47]:
example4.fillna(method='ffill', axis=1)

0    1    2    3
0  1.0  1.0  7.0  7.0
1  2.0  5.0  8.0  8.0
2  NaN  6.0  9.0  9.0

## Exercise (g):

In [48]:
# What output does example4.fillna(method='bfill', axis=1) produce?
# What about example4.fillna(method='ffill') or example4.fillna(method='bfill')?
# Can you think of a longer code snippet to write that can fill all of the null values in example4?
example4.fillna(example4.mean())

0    1  2   3
0  1.0  5.5  7 NaN
1  2.0  5.0  8 NaN
2  1.5  6.0  9 NaN

In [49]:
label = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
label

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

In [50]:
class_labels = {'business class':0,'economy class':1,'first class':2}
label['class'] = label['class'].replace(class_labels)
label

ID  class
0  10      0
1  20      2
2  30      1
3  40      1
4  50      1
5  60      0

In [51]:
one_hot = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
one_hot

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

In [52]:
one_hot_data = pd.get_dummies(one_hot,columns=['class'])
one_hot_data

ID  class_business class  class_economy class  class_first class
0  10                     1                    0                  0
1  20                     0                    0                  1
2  30                     0                    1                  0
3  40                     0                    1                  0
4  50                     0                    1                  0
5  60                     1                    0                  0

### Removing duplicate data 

In [53]:
example6 = pd.DataFrame({'letters': ['A','B'] * 2 + ['B'],
                         'numbers': [1, 2, 1, 3, 3]})
example6

letters  numbers
0       A        1
1       B        2
2       A        1
3       B        3
4       B        3

In [54]:
example6.duplicated()

0    False
1    False
2     True
3    False
4     True
dtype: bool

In [55]:
example6.drop_duplicates()

letters  numbers
0       A        1
1       B        2
3       B        3

In [56]:
example6.drop_duplicates(['letters'])

letters  numbers
0       A        1
1       B        2